src/network/networkd-address.c

```diff
@@ -496,92 +550,69 @@ int address_get(Link *link, const Address *in, Address **ret) {
         return 0;
 }
 
-int link_get_ipv6_address(Link *link, const struct in6_addr *address, Address **ret) {
-        _cleanup_(address_freep) Address *a = NULL;
-        int r;
-
-        assert(link);
-        assert(address);
-
-        r = address_new(&a);
-        if (r < 0)
-                return r;
-
-        /* address_compare_func() only compares the local address for IPv6 case. So, it is enough to
-         * set only family and the address. */
-        a->family = AF_INET6;
-        a->in_addr.in6 = *address;
-
-        return address_get(link, a, ret);
-}
-
-int link_get_ipv4_address(Link *link, const struct in_addr *address, unsigned char prefixlen, Address **ret) {
+int link_get_address(Link *link, int family, const union in_addr_union *address, unsigned char prefixlen, Address **ret) {
+        Address *a;
         int r;
 
         assert(link);
+        assert(IN_SET(family, AF_INET, AF_INET6));
         assert(address);
 
         if (prefixlen != 0) {
-                _cleanup_(address_freep) Address *a = NULL;
+                _cleanup_(address_freep) Address *tmp = NULL;
 
                 /* If prefixlen is set, then we can use address_get(). */
 
-                r = address_new(&a);
+                r = address_new(&tmp);
                 if (r < 0)
                         return r;
 
-                a->family = AF_INET;
-                a->in_addr.in = *address;
-                a->prefixlen = prefixlen;
-
-                return address_get(link, a, ret);
-        } else {
-                Address *a;
-
-                SET_FOREACH(a, link->addresses) {
-                        if (a->family != AF_INET)
-                                continue;
-
-                        if (!in4_addr_equal(&a->in_addr.in, address))
-                                continue;
+                tmp->family = family;
+                tmp->in_addr = *address;
+                tmp->prefixlen = prefixlen;
 
+                if (address_get(link, tmp, &a) >= 0) {
                         if (ret)
                                 *ret = a;
-
                         return 0;
                 }
 
-                return -ENOENT;
+                /* IPv4 addresses may have label and/or broadcast address.
+                 * Hence, we need to always fallback below. */
+                if (family == AF_INET6)
+                        return -ENOENT;
+        }
+
+        SET_FOREACH(a, link->addresses) {
+                if (a->family != family)
+                        continue;
+
+                if (!in_addr_equal(family, &a->in_addr, address))
+                        continue;
+
+                if (in_addr_is_set(family, &a->in_addr_peer))
+                        continue;
+
+                if (ret)
+                        *ret = a;
+
+                return 0;
         }
```

Comment:
Maybe add a comment explaining the logic here? It's not obvious why `in_addr_peer` is checked.